usr.bin/jot/jot.1
.\"	$OpenBSD: jot.1,v 1.24 2021/08/13 07:59:04 martijn Exp $
.\"	$NetBSD: jot.1,v 1.2 1994/11/14 20:27:36 jtc Exp $
.\"
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)jot.1	8.1 (Berkeley) 6/6/93
.\"
.Dd $Mdocdate: August 13 2021 $
.Dt JOT 1
.Os
.Sh NAME
.Nm jot
.Nd print sequential or random data
.Sh SYNOPSIS
.Nm jot
.Bk -words
.Op Fl cnr
.Op Fl b Ar word
.Op Fl p Ar precision
.Op Fl s Ar string
.Op Fl w Ar word
.Oo Ar reps Oo Ar begin Oo Ar end
.Oo Ar step Oc Oc Oc Oc
.Ek
.Sh DESCRIPTION
.Nm
is used to print out increasing, decreasing, random,
or redundant data, usually numbers, one per line.
.Pp
The options are as follows:
.Bl -tag -width "-p precision"
.It Fl b Ar word
Just print
.Ar word
repetitively.
Overrides earlier
.Fl b ,
.Fl c ,
and
.Fl w .
.It Fl c
This is an abbreviation for
.Fl w Ic %c .
Overrides earlier
.Fl b
and
.Fl w .
.It Fl n
Do not print the final newline normally appended to the output.
.It Fl p Ar precision
Print only as many digits or characters of the data
as indicated by the integer
.Ar precision .
In the absence of
.Fl p ,
the precision is the greater of the numbers
.Ar begin
and
.Ar end .
The
.Fl p
option is overridden by whatever appears in a
.Xr printf 3
conversion following
.Fl w .
.It Fl r
Generate random data.
By default,
.Nm
generates sequential data.
.It Fl s Ar string
Print data separated by
.Ar string .
Normally, newlines separate data.
.It Fl w Ar word
Print
.Ar word
with the generated data appended to it.
Octal, hexadecimal, exponential, ASCII, zero-padded,
and right-adjusted representations
are possible by using the appropriate
.Xr printf 3
conversion specification inside
.Ar word ,
in which case the data is inserted rather than appended.
Overrides earlier
.Fl b ,
.Fl c ,
and
.Fl w .
.El
.Pp
The last four arguments specify the length of the output sequence,
its start and end points, and the step size.
Any three of these arguments determine the fourth.
If the given and computed values for
.Ar reps
conflict, the lower value is used.
.Pp
Arguments can be omitted by specifying a
.Ql - .
The default values for
.Ar reps ,
.Ar begin ,
.Ar end ,
and
.Ar step
are 100, 1, 100, and 1, respectively.
Omitted values are computed if possible or assume the default.
A special case arises if only
.Ar begin
and
.Ar end
are specified:
if
.Ar begin
is greater than
.Ar end
then
.Ar step
is set to \(mi1, otherwise it is set to 1;
afterwards
.Ar reps
is computed.
.Pp
.Ar reps
is expected to be an unsigned integer,
and if given as zero is taken to be infinite.
.Ar begin
and
.Ar end
may be given as real numbers or as characters
representing the corresponding value in ASCII.
The last argument must be a real number.
.Pp
Random numbers are obtained through
.Xr arc4random 3 .
Historical versions of
.Nm
used
.Ar step
to seed the random number generator.
This is no longer supported.
The name
.Nm
derives in part from
.Dq iota ,
a function in APL.
.Ss Rounding and truncation
The
.Nm
utility uses double precision floating point arithmetic internally.
Before printing a number, it is converted depending on the output
format used.
.Pp
If no output format is specified or the output format is a
floating point format
.Po
.Sq f ,
.Sq e ,
.Sq g ,
.Sq E ,
or
.Sq G
.Pc ,
the value is rounded using the
.Xr printf 3
function, taking into account the requested precision.
.Pp
If the output format is an integer format
.Po
.Sq c ,
.Sq d ,
.Sq o ,
.Sq x ,
.Sq u ,
.Sq D ,
.Sq O ,
.Sq X ,
.Sq U ,
or
.Sq i
.Pc ,
the value is converted to an integer value by truncation.
.Pp
As an illustration, consider the following command:
.Bd -literal -offset indent
$ jot 6 1 10 0.5
1
2
2
2
3
4
.Ed
.Pp
By requesting an explicit precision of 1, the values generated before rounding
can be seen.
The .5 values are rounded down if the integer part is even,
up otherwise.
.Bd -literal -offset indent
$ jot -p 1 6 1 10 0.5
1.0
1.5
2.0
2.5
3.0
3.5
.Ed
.Pp
By offsetting the values slightly, the values generated by the following
command are always rounded down:
.Bd -literal -offset indent
$ jot -p 0 6 .9999999999 10 0.5
1
1
2
2
3
3
.Ed
.Pp
Another way of achieving the same result is to force truncation by
specifying an integer format:
.Bd -literal -offset indent
$ jot -w %d 6 1 10 0.5
.Ed
.Sh EXAMPLES
Print 21 evenly spaced numbers increasing from \-1 to 1:
.Pp
.Dl $ jot 21 \-1 1.00
.Pp
Generate the ASCII character set:
.Pp
.Dl $ jot \-c 128 0
.Pp
Generate the strings xaa through xaz:
.Pp
.Dl $ jot \-w xa%c 26 a
.Pp
Generate 20 random 8-letter strings:
.Pp
.Dl "$ jot \-r \-c 160 a z | rs \-g0 0 8"
.Pp
Infinitely many
.Xr yes 1 Ns 's
may be obtained through:
.Pp
.Dl $ jot \-b yes 0
.Pp
Thirty
.Xr ed 1
substitution commands applying to lines 2, 7, 12, etc. is the result of:
.Pp
.Dl $ jot \-w %ds/old/new/ 30 2 \- 5
.Pp
Create a file containing exactly 1024 bytes:
.Pp
.Dl $ jot \-b x 512 > block
.Pp
To set tabs four spaces apart starting
from column 10 and ending in column 132, use:
.Pp
.Dl $ expand \-`jot \-s, \- 10 132 4`
.Pp
To print all lines 80 characters or longer:
.Pp
.Dl $ grep `jot \-s \&"\&" \-b . 80`
.Sh SEE ALSO
.Xr ed 1 ,
.Xr expand 1 ,
.Xr rs 1 ,
.Xr yes 1 ,
.Xr arc4random 3 ,
.Xr printf 3
.Sh HISTORY
The
.Nm
utility first appeared in
.Bx 4.2 .
.Sh AUTHORS
.An John A. Kunze


usr.bin/jot/jot.c
/*	$OpenBSD: jot.c,v 1.55 2021/08/13 11:27:33 martijn Exp $	*/
/*	$NetBSD: jot.c,v 1.3 1994/12/02 20:29:43 pk Exp $	*/

/*-
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * jot - print sequential or random data
 *
 * Author:  John Kunze, Office of Comp. Affairs, UCB
 */

#include <ctype.h>
#include <err.h>
#include <limits.h>
#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define	REPS	1
#define	BEGIN	2
#define	ENDER	4
#define	STEP	8

#define	is_default(s)	(strcmp((s), "-") == 0)

static long	reps	= 100;
static double	begin	= 1;
static double	ender	= 100;
static double	step	= 1;

static char	*format = "";
static char	*sepstring = "\n";
static int	prec = -1;
static bool	boring;
static bool	chardata;
static bool	finalnl = true;
static bool	infinity;
static bool	intdata;
static bool	longdata;
static bool	nosign;
static bool	randomize;
static bool	word;

static void	getformat(void);
static int	getprec(char *);
static int	putdata(double, bool);
static void __dead	usage(void);

int
main(int argc, char *argv[])
{
	double		x;
	double		y;
	long		i;
	unsigned int	mask = 0;
	int		n = 0;
	int		ch;
	const char	*errstr;

	if (pledge("stdio", NULL) == -1)
		err(1, "pledge");

	while ((ch = getopt(argc, argv, "b:cnp:rs:w:")) != -1) {
		switch (ch) {
		case 'b':
			boring = true;
			chardata = word = false;
			format = optarg;
			break;
		case 'c':
			chardata = true;
			boring = word = false;
			format = "";
			break;
		case 'n':
			finalnl = false;
			break;
		case 'p':
			prec = strtonum(optarg, 0, INT_MAX, &errstr);
			if (errstr != NULL)
				errx(1, "bad precision value, %s: %s", errstr,
					optarg);
			break;
		case 'r':
			randomize = true;
			break;
		case 's':
			sepstring = optarg;
			break;
		case 'w':
			word = true;
			boring = chardata = false;
			format = optarg;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	switch (argc) {	/* examine args right to left, falling thru cases */
	case 4:
		if (!is_default(argv[3])) {
			if (!sscanf(argv[3], "%lf", &step))
				errx(1, "Bad s value: %s", argv[3]);
			mask |= STEP;
			if (randomize)
				warnx("random seeding not supported");
		}
	case 3:
		if (!is_default(argv[2])) {
			if (!sscanf(argv[2], "%lf", &ender))
				ender = argv[2][strlen(argv[2])-1];
			mask |= ENDER;
			if (prec == -1)
				n = getprec(argv[2]);
		}
	case 2:
		if (!is_default(argv[1])) {
			if (!sscanf(argv[1], "%lf", &begin))
				begin = argv[1][strlen(argv[1])-1];
			mask |= BEGIN;
			if (prec == -1)
				prec = getprec(argv[1]);
			if (n > prec)		/* maximum precision */
				prec = n;
		}
	case 1:
		if (!is_default(argv[0])) {
			reps = strtonum(argv[0], 0, LONG_MAX, &errstr);
			if (errstr != NULL)
				errx(1, "Bad reps value, %s: %s", errstr,
				    argv[0]);
			mask |= REPS;
			if (reps == 0)
				infinity = true;
		}
	case 0:
		if (prec == -1)
			prec = 0;
		break;
	default:
		errx(1, "Too many arguments.  What do you mean by %s?",
		    argv[4]);
	}

	if (!boring)
		getformat();

	if (!randomize) {
		/*
		 * Consolidate the values of reps, begin, ender, step:
		 * The formula ender - begin == (reps - 1) * step shows that any
		 * three determine the fourth (unless reps == 1 or step == 0).
		 * The manual states the following rules:
		 * 1. If four are specified, compare the given and the computed
		 *    value of reps and take the smaller of the two.
		 * 2. If steps was omitted, it takes the default, unless both
		 *    begin and ender were specified.
		 * 3. Assign defaults to omitted values for reps, begin, ender,
		 *    from left to right.
		 */
		switch (mask) { /* Four cases involve both begin and ender. */
		case REPS | BEGIN | ENDER | STEP:
			if (infinity)
				errx(1,
				    "Can't specify end of infinite sequence");
			if (step != 0.0) {
				long t = (ender - begin + step) / step;
				if (t <= 0)
					errx(1, "Impossible stepsize");
				if (t < reps)
					reps = t;
			}
			break;
		case REPS | BEGIN | ENDER:
			if (infinity)
				errx(1,
				    "Can't specify end of infinite sequence");
			if (reps == 1)
				step = 0.0;
			else
				step = (ender - begin) / (reps - 1);
			break;
		case BEGIN | ENDER:
			step = ender > begin ? 1 : -1; /* FreeBSD's behavior. */
			/* FALLTHROUGH */
		case BEGIN | ENDER | STEP:
			if (step == 0.0) {
				reps = 0;
				infinity = true;
				break;
			}
			reps = (ender - begin + step) / step;
			if (reps <= 0)
				errx(1, "Impossible stepsize");
			break;
		case ENDER:		/* Four cases involve only ender. */
		case ENDER | STEP:
		case REPS | ENDER:
		case REPS | ENDER | STEP:
			if (infinity)
				errx(1,
				    "Must specify start of infinite sequence");
			begin = ender - reps * step + step;
			break;
		default:
			/*
			 * The remaining eight cases omit ender.  We don't need
			 * to compute anything because only reps, begin, step
			 * are used for producing output below.  Rules 2. and 3.
			 * together imply that ender will be set last.
			 */
			break;
		}

		for (i = 1, x = begin; i <= reps || infinity; i++, x += step)
			if (putdata(x, reps == i && !infinity))
				errx(1, "range error in conversion: %f", x);
	} else { /* Random output: use defaults for omitted values. */
		bool		use_unif = 0;
		uint32_t	pow10 = 1;
		uint32_t	uintx = 0; /* Initialized to make gcc happy. */

		if (prec > 9)	/* pow(10, prec) > UINT32_MAX */
			errx(1, "requested precision too large");

		if (ender < begin) {
			x = begin;
			begin = ender;
			ender = x;
		}
		x = ender - begin;

		if (prec > 0 || (fmod(ender, 1) == 0 && fmod(begin, 1) == 0)) {
			double range;

			while (prec-- > 0)
				pow10 *= 10;

			range = pow10 * (ender - begin);

			/* If range is an integer, use arc4random_uniform(). */
			if (fmod(range, 1) == 0) {
				if (range >= UINT32_MAX)
					errx(1, "requested range too large");
				use_unif = 1;
				uintx = range + 1;
			}
		}

		for (i = 1; i <= reps || infinity; i++) {
			double v;

			if (use_unif) {
				y = arc4random_uniform(uintx) / (double)pow10;
				v = y + begin;
			} else {
				y = arc4random() / ((double)0xffffffff + 1);
				v = y * x + begin;
			}
			if (putdata(v, reps == i && !infinity))
				errx(1, "range error in conversion: %f", v);
		}
	}

	if (finalnl)
		putchar('\n');

	return 0;
}

static int
putdata(double x, bool last)
{
	if (boring)
		printf("%s", format);
	else if (longdata && nosign) {
		if (x < 0.0 || x > (double)ULONG_MAX)
			return 1;
		printf(format, (unsigned long)x);
	} else if (longdata) {
		if (x < (double)LONG_MIN || x > (double)LONG_MAX)
			return 1;
		printf(format, (long)x);
	} else if (chardata || (intdata && !nosign)) {
		if (x < (double)INT_MIN || x > (double)INT_MAX)
			return 1;
		printf(format, (int)x);
	} else if (intdata) {
		if (x < 0.0 || x > (double)UINT_MAX)
			return 1;
		printf(format, (unsigned int)x);
	} else
		printf(format, x);
	if (!last)
		fputs(sepstring, stdout);

	return 0;
}

static void __dead
usage(void)
{
	(void)fprintf(stderr, "usage: jot [-cnr] [-b word] [-p precision] "
	    "[-s string] [-w word]\n"
	    "	   [reps [begin [end [step]]]]\n");
	exit(1);
}

static int
getprec(char *s)
{
	if ((s = strchr(s, '.')) == NULL)
		return 0;
	return strspn(s + 1, "0123456789");
}

static void
getformat(void)
{
	char	*p;

	p = format;
	while ((p = strchr(p, '%')) != NULL && p[1] == '%')
		p += 2;

	if (p == NULL && !chardata) {
		if (asprintf(&format, "%s%%.%df", format, prec) == -1)
			err(1, NULL);
	} else if (p == NULL && chardata) {
		if (asprintf(&format, "%s%%c", format) == -1)
			err(1, NULL);
	} else if (p[1] == '\0') {
		/* cannot end in single '%' */
		if (asprintf(&format, "%s%%", format) == -1)
			err(1, NULL);
	} else {
		/*
		 * Allow conversion format specifiers of the form
		 * %[#][ ][{+,-}][0-9]*[.[0-9]*]? where ? must be one of
		 * [l]{d,i,o,u,x} or {f,e,g,F,E,G,d,o,x,D,O,U,X,c,u}
		 */
		char	*fmt;
		int	dot, hash, space, sign, numbers;

		fmt = p++;
		dot = hash = space = sign = numbers = 0;
		while (!isalpha((unsigned char)*p)) {
			if (isdigit((unsigned char)*p)) {
				numbers++;
				p++;
			} else if ((*p == '#' && !(numbers|dot|sign|space|
			    hash++)) ||
			    (*p == ' ' && !(numbers|dot|space++)) ||
			    ((*p == '+' || *p == '-') && !(numbers|dot|sign++))
			    || (*p == '.' && !(dot++)))
				p++;
			else
				goto fmt_broken;
		}
		if (*p == 'l') {
			longdata = true;
			if (*++p == 'l') {
				p++;
				goto fmt_broken;
			}
		}
		switch (*p) {
		case 'd':
		case 'i':
			intdata = true;
			break;
		case 'o':
		case 'u':
		case 'x':
		case 'X':
			intdata = nosign = true;
			break;
		case 'D':
			if (longdata)
				goto fmt_broken;
			longdata = intdata = true; /* same as %ld */
			break;
		case 'O':
		case 'U':
			if (longdata)
				goto fmt_broken;
			longdata = intdata = nosign = true; /* same as %l[ou] */
			break;
		case 'c':
			if (longdata)
				goto fmt_broken;
			chardata = true;
			break;
		case 'e':
		case 'E':
		case 'f':
		case 'F':
		case 'g':
		case 'G':
			if (longdata)
				goto fmt_broken;
			/* No cast needed for printing in putdata() */
			break;
		default:
fmt_broken:
			errx(1, "illegal or unsupported format '%.*s'",
			    (int)(p + 1 - fmt), fmt);
		}

		while ((p = strchr(p, '%')) != NULL && p[1] == '%')
			p += 2;

		if (p != NULL) {
			if (p[1] != '\0')
				errx(1, "too many conversions");
			/* cannot end in single '%' */
			if (asprintf(&format, "%s%%", format) == -1)
				err(1, NULL);
		}
	}
}
